sklearn/impute/_base.py

```diff
@@ -437,20 +441,24 @@ def transform(self, X):
                                  "array instead.")
             else:
                 mask = _get_mask(X.data, self.missing_values)
+                _, missing_mask = _get_mask(X_, self.missing_values, True)
```

Comment:
I don't think there is any benefit in getting both `mask` and `missing_mask`. Why do we want that duplication? and please choose variable names that allow the reader to understand the distinction between them.

Reply:
If you look closely `mask` is being found using `X.data` whereas `missing_mask` is found using `X_`, if you see the code just before it, we are removing the features that are invalid. But since here we will use `transform` function and we fit out `MissingIndicator` with the entire `missing_mask` of `X` without eliminating the invalid features, therefore the duplicacy or the test cases would give an error stating `X has different number of features while fitting`

Reply:
I did fix some the issues you mentione in this review and also stated why some of them can't be fixed, if you find me incorrect kindly comment. I am not pushing a commit until then.

Thanks.

Reply:
Ah... Pesky invalid features. So if I'm understanding correctly, X_ does not have invalid features removed, but X does at this point. And at transform time we pass the mask of the full X, not removing invalid features, to be transformed by the MissingIndicator. Since invalid features is a rare case, I'd rather not make things especially expensive in terms of code or computation, for their sake. Could we instead remove the invalid features from the mask of X at transform time??

Reply:
Yup, you got that right. Instead of removing the invalid features from invalid mask, we can first transform the MissingIndicator, then remove find and remove the invalid features. How does it sound?

Thanks.

Reply:
>  Instead of removing the invalid features from invalid mask, we can first transform the MissingIndicator, then find and remove the invalid features. How does it sound?

Doesn't that still involve maintaining a separate X and X_ at training time?

The whole purpose of the present pull request is to avoid duplicated work, especially extracting the mask from X twice. Your code does just that, by processing X and then X_. So we need to find an alternative that does not do that, or else this pull request does not help us.

Reply:
We could also consider a solution that was neat when there were no invalid features, and that duplicated work if there were invalid features.

Reply:
I have removed all-1 duplicacies including `X_`, the one that remains is the call to `_get_mask` when invalid indices are eliminated. I think I can remove that too, working on it.
Have also optimized `_get_mask` (removed the duplicacy of `X` to `X_`). Kindly review and let me know if something needs to be changed.

Thanks.